internal/ls/signaturehelp.go

```diff
@@ -0,0 +1,1120 @@
+	return l.GetSignatureHelpItems(fileName, position, program, sourceFile, context, clientOptions, preferences)
+}
+
+func (l *LanguageService) GetSignatureHelpItems(
+	fileName string,
+	position int,
+	program *compiler.Program,
+	sourceFile *ast.SourceFile,
+	context *lsproto.SignatureHelpContext,
+	clientOptions *lsproto.SignatureHelpClientCapabilities,
+	preferences *UserPreferences) *lsproto.SignatureHelp {
+	typeChecker := program.GetTypeChecker()
+
+	// Decide whether to show signature help
+	startingToken := astnav.FindPrecedingToken(sourceFile, position)
+	if startingToken == nil {
+		// We are at the beginning of the file
+		return nil
+	}
+
+	// Only need to be careful if the user typed a character and signature help wasn't showing.
+	onlyUseSyntacticOwners := context.TriggerKind == 2
+
+	// Bail out quickly in the middle of a string or comment, don't provide signature help unless the user explicitly requested it.
+	if onlyUseSyntacticOwners && IsInString(sourceFile, position, startingToken) { // isInComment(sourceFile, position) needs formatting implemented
+		return nil
+	}
+
+	isManuallyInvoked := context.TriggerKind == 1
+	argumentInfo := getContainingArgumentInfo(startingToken, sourceFile, program.GetTypeChecker(), isManuallyInvoked, position)
+	if argumentInfo == nil {
+		return nil
+	}
+
+	// cancellationToken.throwIfCancellationRequested();
+
+	// Extra syntactic and semantic filtering of signature help
+	candidateInfo := getCandidateOrTypeInfo(argumentInfo, typeChecker, sourceFile, startingToken, onlyUseSyntacticOwners)
+	// cancellationToken.throwIfCancellationRequested();
+
+	// if (!candidateInfo) { !!!
+	// 	// We didn't have any sig help items produced by the TS compiler.  If this is a JS
+	// 	// file, then see if we can figure out anything better.
+	// 	return isSourceFileJS(sourceFile) ? createJSSignatureHelpItems(argumentInfo, program, cancellationToken) : undefined;
+	// }
+
+	// return typeChecker.runWithCancellationToken(cancellationToken, typeChecker =>
+	if candidateInfo.candidateInfo != nil {
+		return createSignatureHelpItems(candidateInfo.candidateInfo.candidates, candidateInfo.candidateInfo.resolvedSignature, argumentInfo, sourceFile, typeChecker, onlyUseSyntacticOwners, clientOptions)
+	}
+	return createTypeHelpItems(candidateInfo.typeInfo, argumentInfo, sourceFile, clientOptions, typeChecker)
+}
+
+func createTypeHelpItems(symbol *ast.Symbol, argumentInfo *argumentListInfo, sourceFile *ast.SourceFile, clientOptions *lsproto.SignatureHelpClientCapabilities, c *checker.Checker) *lsproto.SignatureHelp {
+	typeParameters := c.GetLocalTypeParametersOfClassOrInterfaceOrTypeAlias(symbol)
+	if typeParameters == nil {
+		return nil
+	}
+	item := getTypeHelpItem(symbol, typeParameters, getEnclosingDeclarationFromInvocation(argumentInfo.invocation), sourceFile, c)
+
+	// Converting signatureHelpParameter to *lsproto.ParameterInformation
+	var parameters []*lsproto.ParameterInformation
```

Comment:
Some of these functions are converting from local types to lsproto type because they needed some extra parameters like `isRest`, `isOptional` or `isVariadic` to be able to correctly calculate the parameter count and the parameter index that would be highlighted.